# Proyek Analisis Data: [Bike-sharing-dataset]
- **Nama:** [Yuanita TriHastutik Rachmawati]
- **Email:** [yuanitatri987@gmail.com]
- **ID Dicoding:** [yuanita_9dre]

## Menentukan Pertanyaan Bisnis

1. Kapan dan pada jam berapa peminjaman sepeda melonjak tinggi pada tahun 2011??
2. Apakah terdapat perbedaan pelonjakan yang terjadi di tahun 2011 dan 2012?

## Import Semua Packages/Library yang Digunakan

Penggunaan library yang didukung oleh python,
- pandas untuk membaca dataset yang berupa dataset yang bertype csv
- numpy untuk perhitungan matematika

In [ ]:
import pandas as pd
import numpy as np
import matplotlib  as plt
import seaborn as sns

## Data Wrangling

### Gathering Data

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# # #data menurut hari
dt_day= pd.read_csv('/content/drive/MyDrive/Bike-sharing-dataset/day.csv') #untuk data peminjaman sepeda
dt_day


Memuat data peminjaman sepeda berdasarkan parameter yang sudah disediakan. terdapat 730 data dan 16 kolom parameter

In [ ]:
# Load data menurut jam
dt_hour= pd.read_csv('/content/drive/MyDrive/Bike-sharing-dataset/hour.csv') #untuk data peminjaman sepeda
dt_hour

Memuat data peminjaman sepeda berdasarkan setaio jam dalam satu hari. terdapat 17378 row dan 17 kolom

### Assessing Data

**Menilai data dt_day**

In [ ]:
dt_day.info()

In [ ]:
dt_day.isnull().sum()

In [ ]:
print("Jumlah duplikasi: ",dt_day.duplicated().sum())
dt_day.describe()

Setelah dilakukan penilaian tidak terdapat missing value pada data hari, namun type data pada kolom dteday yang seharusnya bertype data datetime namun disini bertype data object, karena dateday merupakan data tanggal seperti tahun-bulan-tanggal

**Menilai data dt_hour**

In [ ]:
dt_hour.info()

In [ ]:
dt_hour.isna().sum()

In [ ]:
print("Jumlah duplikasi: ",dt_hour.duplicated().sum())
dt_hour.describe()

Setelah dilakukan assesing pada data jam tidak terdapat missing value atau duplikasi yang ada, namun terdapat kesalahan dalam tipe data untuk kolom dteday

**Insight:**
- Dalam rekap data hari tidak ditemukan missing valaue, dicek melalui dt_day.info(),terdapat 15 kolom dan 731 row, 16 type data.
- Dalam rekap data peminjaman per-jam tidak ditemukan juga missing value,dicek melalui dt_hour.info() terdapat 16 kolom dengan 17379 row, 17 type data.
- Namun terdapat kesalahan typedata diaman dteday seharusnya bertipe data datetime buka sebagai object, karena object diguanakan untuk string

### Cleaning Data

Cleaning data dilakukan apabila terdapat data yang tidak akurat,missing value, pada dataset ini tidak terdapat missing value ataupun data yang duplikat. Namun saya mengubah type "object" dari kolom dteday ke type data datetime. Kemudian hal yang sama juga saya lakukan pada data_hour, karena terdapat kesalah type data pada dteday peminjaman.

1. Define : dt_day memiliki parameter instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt. Dengan target yang akan saya  analisis adalah mencari tahu penyebab minat peminjam sepeda sesuai dengan kategori hari/tahun. Karena terdapat kesalahan type data maka akan dilakukan perubahan type data terlebih dahulu.
2. Code : Eksekusi program
3. Test : Menjalankan kode program

In [ ]:
dt_day["dteday"] = pd.to_datetime(dt_day['dteday'], format='%Y-%m-%d')
dt_day.info()

In [ ]:
dt_hour["dteday"] = pd.to_datetime(dt_hour['dteday'], format='%Y-%m-%d')
dt_hour.info()

In [ ]:
dt_day['yr'] = dt_day['yr'].replace({0:2011, 1: 2012})
dt_day

In [ ]:
dt_hour['yr'] = dt_hour['yr'].replace({0:2011, 1: 2012})

**Insight:**
- Perubahan dilakukan pada kedua data memanfaatkan method dari library pandas yaitu pd.to_datetime dengan format year-month-day.
- Pada saat dicek kembali menggunakan dt_hour.info() dteday sudah berganti typedata dengan datetime64[ns]

## Exploratory Data Analysis (EDA)

### Explore ...

Explorasi saya lakukan dengan melakukan beberapa penggabungan data dari dataset yang tesedia, melakukan groupby dengan target yang sudah ditentukan

In [ ]:
dt_day.describe(include="all")

In [ ]:
dt_day.groupby(by="weathersit").agg({
    "instant" : "nunique",
    "registered" :["max", "min", "mean", "std"]
})

In [ ]:
dt_day.groupby(by="workingday").agg({"registered" :["max", "min", "mean", "std"]})

In [ ]:
dt_day.groupby(by="season").instant.nunique().sort_values(ascending=True)

In [ ]:
dt_day.groupby(by="yr").casual.nunique().sort_values(ascending=False)

Dari pivot table diatas, pengelompokan data peminjaman sepeda berdasarkan musim. Melalui daftar pelanggan yang sudah register pada musim pertama(cuaca cerah) sebanyak 6946 dan paling sedikit 416, pada musim kedua sebanyak 6844 dan paling sedikit 491, pada musim ketiga terdapat penurunan 4324 dan paling sedikit 20 peminjam.

**Insight:**
- Perubahan saya lakukan pada dt_day mengganti kategori taun 2011=0, 2012=1 menggunakan method replace

## Visualization & Explanatory Analysis

### Pertanyaan 1:
1. Kapan dan pada jam berapa peminjaman sepeda melonjak tinggi pada tahun 2011??


**DATASET = DT_DAY**

In [ ]:
by_2011 = dt_day[dt_day['yr'] == 2011]
hasil=by_2011.groupby(by="dteday")['cnt'].sum()
sorted_hasil = hasil.sort_values(ascending=False)
sorted_hasil #dihasilkan pada tanggal 2011-07-04 dengan jumlah 6043


In [21]:
result = '2011-07-04'
result2 =dt_day[dt_day['dteday']==result]
result2

instant     dteday  season    yr  mnth  holiday  weekday  workingday  \
184      185 2011-07-04       3  2011     7        1        1           0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
184           2  0.726667  0.665417  0.637917   0.081479    3065        2978   

      cnt  
184  6043

**PENJELASAN**

Variabel by_2011 menyimpan data khusus untuk 2011. data tersebut digroup dengan kolom count yang disimpan dalam variabel hasil. Setelah data menjadi group kemudian di Descanding agar mengetahui mana data yang memiliki nilai max. data max tahun 2011 tercatat pada 2011-07-04 dengan jumlah 6043 data. Data yang didapat pada tanggal tesebut mencoba dikelompokan dengan dteday karena data pada tanggal tersebut masih terpencar. Kemudian terdapat variabel result untuk menyimpan data tanggal yang akan dilihat datanya yaitu 2011-07-04. Variabel result2 digunakan untuk mengambil data yang sama dari kolom dteday sesuai dengan result.


**DATASET = DT_HOURS**

In [ ]:
by_2011 = dt_hour[dt_hour['yr'] == 2011]
hasil=by_2011.groupby(by="dteday")['cnt'].sum()
hasil

In [ ]:
result = '2011-07-04'
result2 =dt_hour[dt_hour['dteday']==result]
result2



In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

plt.figure(figsize=(15, 10))
colors=['#72BCD4']
sns.barplot(
    x='hr',
    y='cnt',
    data=result2, palette=colors)
plt.title('Peminjaman Sepeda pada 2011-07-04 berdasarkan Jam', fontsize=16)
plt.xlabel('Jam', fontsize=14)
plt.ylim(0,500)
plt.ylabel('Jumlah Peminjaman Sepeda', fontsize=14)
plt.show()



### Pertanyaan 2:

2. Apakah terdapat perbedaan pelonjakan yang terjadi di tahun 2011 dan 2012?

In [ ]:
dt_day['yr'] = dt_day['yr'].replace({0:2011, 1: 2012})

In [ ]:
group_df = dt_day.groupby('yr')['cnt'].sum().reset_index()
group_df

yr      cnt
0  2011  1243103
1  2012  2049576

In [ ]:
import matplotlib.pyplot as plt

plt.figure(figsize=(12, 6))
sns.barplot(data=group_df, x='yr', y='cnt', hue='yr', palette='Set1')
plt.title('Perbandingan Peminjaman Sepeda antara Tahun 2011 dan 2012', fontsize=16)
plt.xlabel('Tahun', fontsize=14)
plt.ylabel('Jumlah Peminjaman Sepeda', fontsize=14)
plt.legend(title='Tahun')
plt.show


**Insight:**
- Pertanyaan 1 : Dari analisis dataset dt_day dan dt_hour, didapatkan peminjaman sepeda mengalami pelonjakan paling tinggi pada 2011-07-04 dengan jumlah peminjam yaitu 6034 data yang terdiri dari 2 kategori peminjam yaitu register dan casual. Dalam 23 Jam peminjaman paling tinggi terjadi pada jam 21 dengan jumlah 457 peminjam. Visualisasi data menggunakan barplot dengan sumbu x adalah Waktu: Jam, dan sumbu y:jumlah peminjam

- Pertanyaan 2: Perbedaan pelonjakan peminjam sepeda pada tahun 2011 dan 2012 sangat signifikan pada 2011 jumlah data peminjam 1243103, dan 2012 mengalami pelonjakan yaitu 2049576 data peminjam.

In [27]:
all_data = pd.concat([dt_day, dt_hour], ignore_index=True)
all_data.to_csv("all_data.csv", index=False)

## Analisis Lanjutan (Opsional)

## Conclusion

- Dari analisis data untuk menjawab pertanya pertama dapa disimpulkan terdapat kuantitas paling tinggi peminjaman sepeda pada jam dan hari hari tertent. Dari seluruh data di tahun 2011 terdapat 1 hari dimana kuantitas peminjam adalah paling tinggi yaitu 2011-07-04 dengan jumlah peminjam yaitu 457 orang dan peeminjaman paling tinggi terjadi pada jam 21.  Terdapat berbagai parameter yang bisa saya analisis, seperti tanggal,bulan,tahun,musim, ketegori peminjam termasuk dalam casual atau register.
- Analisis data kedua  untuk mengetahui perbandingan kuantitas peminjam di tahun 2011 dan 2012. Pada tahun 2012 peminjam mencapai 2049576 lebih banyak dari 2011, karena pada 2011 peminjam hanya 1243103.  Perbedaan mencapai 806.473 orang.